README.md
README
======

Parsifal is an OCaml-based parsing engine.

Parsifal is a collection of binary parsers and tools. The development
is at an early stage (which explains the 0.1 version).

There are several file formats or network protocols currently (at
least partially) described:

  * X.509 certificates
  * SSL/TLS messages
  * DNS messages
  * MRT/BGP messages
  * Portable Executables
  * UEFI Firmwares
  * PKCS#1 keys and containers
  * PKCS#7 containers
  * Kerberos messages
  * OpenPGP messages
  * DVI documents
  * PNG images
  * PCAP/IP/TCP/UDP rudimentary support
  * NTP messages


Here is the content of the various directories of parsifal repository:

  * syntax/ contains the preprocessor used to generate automatically
    types and functions
  * core/ is the standard parsifal library (common PTypes, input
    structures, useful functions to print values)
  * crypto/ contains the cryptographic functions and object
    descriptions:
    * hash function (MD5, SHA1 and SHA256)
    * Diffie-Hellman keys
    * DSA keys
    * RSA keys and implementation (PKCS#1)
    * PRNG
    * X.509 certificates
    * PKCS#7 containers
  * net/ describes some formats/protcols related to networking
    * PCAP/IP/TCP/UDP trivial support
    * BGP/MRT messages
    * DNS messages
    * NTP messages
  * ssl/ is a first step towards a functionnal TLS stack. For the
    moment, it contains the description of handshake messages and some
    useful functions to produce and read TLS records.
  * formats/ describes some file formats
    * DVI (DeVice Independant files)
    * Portable Executable
    * UEFI Firmware Volumes
    * TAR archives
    * PNG images
  * kerby/ is a collection of files to parse Kerberos messages
  * pci/ is about PCI Expansion ROMs

  * ssl-tools/ contains SSL/TLS useful programs
  * openpgp-tools/ contains a program to parse PGP containers
  * tools/ contains several tools like asn1parse or parsifal, which
    allows to parse and explore described PTypes

Moreover, several test/ and unit/ exist, that contain unfinished
programs and unit tests. usrlibocaml/ is only there to ease the
compilation process.

Finally, tutorial/ and papers/ contain the documentation and submitted
papers describing parsifal, whereas docs contains RFCs and official
specs.


Changelog
2016-12-30 Olivier Levillain <olivier.levillain@ssi.gouv.fr>
	* Various bugfixes to make parsifal and concerto work
	  on recent versions of the OCaml compiler.

2015-07-19 Baptiste Gourdin <baptiste.gourdin@gmail.com>
	* New CRL PType.
	* First version of x509check tool (not commited
	  in master branch, integration pending).

2014-11-07 Nicolas Vivet <nicolas.vivet@ssi.gouv.fr>
	* Bugfix: correctly handle kerberos/heimdal on various
	  systems.

2014-11-05 Olivier Levillain <olivier.levillain@ssi.gouv.fr>
	* Pkcs1.raw_verify security fix
	* Various bugfixes/improvements in tools
	* Bugfix: probe_server probe2dump leaked file
	  descriptors.

2014-08-19 Pierre Chambart <pierre.chambart@ocamlpro.com>
	* Add first support for OPAM

2014-08-13 Olivier Levillain <olivier.levillain@ssi.gouv.fr>
	* Various crypto bugfixes
	* Add first support for record protection
	* Full rewrite of TlsEngineNG to handle Handshake
	  up until Finished

2014-07-18 Olivier Levillain <olivier.levillain@ssi.gouv.fr>
	* Add name extraction to some tools (mapAnswers and
	  x509show)

2014-06-03 Thomas Calderon <thomas.calderon@ssi.gouv.fr>
	* Add support for Keberos encrypted containers using
	  libkrb5
	* First support for PK_INIT RSA decryption
	* Add support for ticket decryption in AP_REQ messages
	* Add support for some of Kerberos authorization data
	  (PAC in particular)

2014-04-04 Florian Maury <florian.maury@ssi.gouv.fr>
	* First description of NTP messages
	* Add function to read input with read_line
	  (1 line = 1 message)
	* Add output option to allow one-line JSON output format

2014-03-10 Olivier Levillain <olivier.levillain@ssi.gouv.fr>
	* Fixed some bugs in TlsEngineNG and how TLS records
	  were parsed offline.

2014-02-27 Olivier Levillain <olivier.levillain@ssi.gouv.fr>
	* New step-by-step example: PKCS#10 CSR

2014-02-25 Olivier Levillain <olivier.levillain@ssi.gouv.fr>
	* Major rewrite of TLS engine (using automata)
	* Remove several programs to limit the number of targets
	* New library: crypto, to factor some code
	* Issue #6: Document the repo layout in README.md
	* Issue #20: Remove Lwt from most of the libs (only the
	  ssl libs and some progs still use Lwt)

2014-02-17 Olivier Levillain <olivier.levillain@ssi.gouv.fr>
	* First implementation of gzip_container
	* New PTypes: (rtol_)bit_magic

2014-02-12 Olivier Levillain <olivier.levillain@ssi.gouv.fr>
	* Issue #15: Add string_input_of_stdin
	* Issue #13: Debug feature: handle name in containers
	* Issue #18 and #19: Add README and INSTALL files
	* First unit tests for the preprocessor
	* Remove Lwt as a specific input type

2014-01-22 Pierre Chifflier <pierre.chifflier@ssi.gouv.fr>
	* Various bugfixes/enhancements
	* PE support improvements
	* First description of UEFI Volume formats

2014-01-21 Florian Maury <florian.maury@ssi.gouv.fr>
	* First description of OpenPGP messages

2014-01-20 Olivier Levillain <olivier.levillain@ssi.gouv.fr>
	* New helper in core: string_input_of_stdin

2014-01-20 Olivier Levillain <olivier.levillain@ssi.gouv.fr>
	* Issue #14: Improve debug info on parsing errors
	  (the name used for a container is now the name of
	  the struct field or the PType [union/alias])

2014-01-15 Olivier Levillain <olivier.levillain@ssi.gouv.fr>
	* Issue #11: Allow for more flexible enum patterns
          (multiple and range patterns)

2014-01-11 Olivier Levillain <olivier.levillain@ssi.gouv.fr>
	* Initial Changelog


x509.ml
open Parsifal
open PTypes
open Asn1Engine
open Asn1PTypes
open X509Basics
open X509Extensions


(***********************)
(* tbs and Certificate *)
(***********************)

asn1_alias issuerUniqueId = primitive [C_ContextSpecific, 1] der_bitstring
asn1_alias subjectUniqueId = primitive [C_ContextSpecific, 2] der_bitstring

struct tbsCertificate_content = {
  optional version : asn1 [(C_ContextSpecific, true, T_Unknown 0)] of der_smallint;
  serialNumber : der_integer;
  signature : algorithmIdentifier;
  parse_checkpoint position_before_issuer : save_offset;
  issuer : distinguishedName;
  parse_field issuer_raw : raw_value(position_before_issuer);
  validity : validity;
  parse_checkpoint position_before_subject : save_offset;
  subject : distinguishedName;
  parse_field subject_raw : raw_value(position_before_subject);
  subjectPublicKeyInfo : subjectPublicKeyInfo;
  optional issuerUniqueId : asn1 [(C_ContextSpecific, false, T_Unknown 1)] of der_bitstring_content;
  optional subjectUniqueId : asn1 [(C_ContextSpecific, false, T_Unknown 2)] of der_bitstring_content;
  optional extensions : asn1 [(C_ContextSpecific, true, T_Unknown 3)] of extension_list
}
asn1_alias tbsCertificate

struct certificate_content = {
  parse_checkpoint position_before_tbs : save_offset;
  tbsCertificate : tbsCertificate;
  parse_field tbsCertificate_raw : raw_value(position_before_tbs);
  signatureAlgorithm : algorithmIdentifier;
  parse_checkpoint _constraint : both_equal(false; (CustomException "signature algos should be equal");
                                            tbsCertificate.signature; signatureAlgorithm);
  signatureValue : bitstring_container of signature(signatureType_of_algo signatureAlgorithm)
}
asn1_alias certificate [top]




(**************************)
(* Populating directories *)
(**************************)

let attribute_value_types = [
  [85; 4; 41], "name", None, AVT_DirectoryString(AtMost 32768);
  [85; 4; 4], "surname", None, AVT_DirectoryString(AtMost 32768);
  [85; 4; 42], "givenName", None, AVT_DirectoryString(AtMost 32768);
  [85; 4; 43], "initials", None, AVT_DirectoryString(AtMost 32768);
  [85; 4; 44], "generationQualifier", None, AVT_DirectoryString(AtMost 32768);

  [85; 4; 3], "commonName", Some "CN", AVT_DirectoryString(AtMost 64);
  [85; 4; 7], "localityName", Some "L", AVT_DirectoryString(AtMost 128);
  [85; 4; 8], "stateOrProvinceName", Some "S", AVT_DirectoryString(AtMost 128);
  [85; 4; 10], "organizationName", Some "O", AVT_DirectoryString(AtMost 64);
  [85; 4; 11], "organizationalUnitName", Some "OU", AVT_DirectoryString(AtMost 64);
  [85; 4; 12], "title", None, AVT_DirectoryString(AtMost 64);
  [85; 4; 46], "dnQualifier", None, AVT_PrintableString(NoConstraint);
  [85; 4; 6], "countryName", Some "C", AVT_PrintableString(AtMost 2);
  [85; 4; 5], "serialNumber", Some "SN", AVT_PrintableString(AtMost 64);
  [85; 4; 65], "pseudonym", None, AVT_DirectoryString(AtMost 128);

  [9; 2342; 19200300; 100; 1; 25], "domainComponent", Some "dc", AVT_IA5String(NoConstraint);
  [42;840;113549;1;9;1], "emailAddress", None, AVT_IA5String(AtMost 255)
]


let dsa_spk_of_param = function
  | Some (DSAParams dp) -> SPK_DSA dp
  | _ -> SPK_Unknown

let public_key_types = [
  [42;840;113549;1;1;1], "rsaEncryption", APT_Null, (fun _ -> SPK_RSA);
  [42;840;10040;4;1], "dsa", APT_DSAParams, dsa_spk_of_param
]

let signature_types = [
  [42;840;113549;1;1;2], "md2WithRSAEncryption", APT_Null, (fun _ -> ST_RSA);
  [42;840;113549;1;1;3], "md4WithRSAEncryption", APT_Null, (fun _ -> ST_RSA);
  [42;840;113549;1;1;4], "md5WithRSAEncryption", APT_Null, (fun _ -> ST_RSA);
  [42;840;113549;1;1;5], "sha1WithRSAEncryption", APT_Null, (fun _ -> ST_RSA);
  [42;840;113549;1;1;11], "sha256WithRSAEncryption", APT_Null, (fun _ -> ST_RSA);
  [42;840;113549;1;1;12], "sha384WithRSAEncryption", APT_Null, (fun _ -> ST_RSA);
  [42;840;113549;1;1;13], "sha512WithRSAEncryption", APT_Null, (fun _ -> ST_RSA);
  [42;840;113549;1;1;14], "sha224WithRSAEncryption", APT_Null, (fun _ -> ST_RSA);
  [42;840;10040;4;3], "dsaWithSha1", APT_Null, (fun _ -> ST_DSA);
  [96;840;1;101;3;4;3;1], "dsaWithSha224", APT_Null, (fun _ -> ST_DSA);
  [96;840;1;101;3;4;3;2], "dsaWithSha256", APT_Null, (fun _ -> ST_DSA);
]

let extension_types = [
  [85;29;1], "authorityKeyIdentifier";
  [85;29;14], "subjectKeyIdentifier";
  [85;29;15], "keyUsage";
  [85;29;17], "subjectAlternativeName";
  [85;29;19], "basicConstraints";
  [85;29;30], "nameConstraints";
  [85;29;31], "crlDistributionPoints";
  [85;29;32], "certificatePolicies";
  [85;29;35], "authorityKeyIdentifier";
  [85;29;37], "extendedKeyUsage";
  [43;6;1;5;5;7;1;1], "authorityInfoAccess";
]


let other_oids = [
  (* Prefixes *)
  [43;6;1;5;5;7], "id-pkix";
  [43;6;1;5;5;7;1], "id-pe";
  [43;6;1;5;5;7;2], "id-qt";
  [43;6;1;5;5;7;3], "id-kp";
  [43;6;1;5;5;7;48], "id-ad";
  [85;29], "id-ce";

  [85;29;37;0], "anyExtendedKeyUsage";
  [43;6;1;5;5;7;3;1], "serverAuth";
  [43;6;1;5;5;7;3;2], "clientAuth";
  [43;6;1;5;5;7;3;3], "codeSigning";
  [43;6;1;5;5;7;3;4], "emailProtection";
  [43;6;1;5;5;7;3;8], "timeStamping";
  [43;6;1;5;5;7;3;9], "OCSPSigning";

  [85;29;32;0], "anyPolicy";
  [43;6;1;5;5;7;2;1], "id-qt-cps";
  [43;6;1;5;5;7;2;2], "id-qt-unotice";

  [43;6;1;5;5;7;48;1], "id-ad-ocsp";
  [43;6;1;5;5;7;48;2], "id-ad-caIssuers";
]


let _ =
  List.iter (populate_atv_directory) attribute_value_types;
  List.iter (fun (id, name) -> register_oid id name) extension_types;
  List.iter (fun (id, name) -> register_oid id name) other_oids;
  List.iter (populate_alg_directory subjectPublicKeyType_directory) public_key_types;
  List.iter (populate_alg_directory signatureType_directory) signature_types;  
  ()


